### PR TITLE
Add searchable trait to checkbox list component

### PR DESCRIPTION
**Pull request by johncarter-**
Allows checkbox list component to be searchable:

**Notes**

`CanBeSearchable` trait property `$searchingMessage` is redundant in this situation; unless you want me to add a delay to results being revealed?

New Recording - 21_11_2022, 17_17_40-high

## Todo

- [x] Disabled state
- [x] Select/deselect all search results

### CODE REVIEW

**Review by zepfietje — CHANGES_REQUESTED** [review]
Could you fix the Blade indentation? It currently contains a lot of unrelated changes.

**Comment by johncarter-** [comment]
Oh man 😢 - I hoped you had a GitHub action that normalised that. I can try and get it back inline with the repo's style.

I use VS Code and the extensions in the screenshot, do you know if there's any settings / ways to keep the styling the same?

image

**Comment by zepfietje** [comment]
I'm also using VS Code, but don't have a formatter for Blade. None of them really gets it all right. Unfortunately there's no Prettier plugin yet. Most other parts of the codebase do get formatted automatically though (PHP using Pint and CSS/JS with Prettier).

**Comment by johncarter-** [comment]
Is that any better?

**Comment by zepfietje** [comment]
Not entirely yet. Will have a look.

**Comment by johncarter-** [comment]
Sorry, bit of a PR noob.

File: packages/forms/resources/views/components/checkbox-list.blade.php [hunk 1]
```diff
@@ -101,7 +149,17 @@
             @endforeach
         </x-filament-support::grid>
 
-    @if ($isBulkToggleable())
+        @if ($isSearchable())
+            <div
+                x-cloak
+                x-show="visibleCheckboxListItems.length === 0"
+                class="text-sm filament-forms-checkbox-list-component-no-search-results-message"
```
Comment (zepfietje) [inline packages/forms/resources/views/components/checkbox-list.blade.php]:
```suggestion
                class="filament-forms-checkbox-list-component-no-search-results-message text-sm"
```

**Review by zepfietje — CHANGES_REQUESTED** [review]

**Comment by danharrin** [comment]
@zepfietje is there anything left to do here before I review?

**Comment by zepfietje** [comment]
Haven't yet had a look at @johncarter-'s latest changes.
Feel free to review, but let me do a final review as well since this PR is quite tricky.

**Comment by danharrin** [comment]
Now we merged in a PR that created many merge conflicts. Its best to hide the select all / deselect all buttons when a search query is active IMO, instead of dealing with "visible" checkboxes.

**Comment by johncarter-** [comment]
> Now we merged in a PR that created many merge conflicts. Its best to hide the select all / deselect all buttons when a search query is active IMO, instead of dealing with "visible" checkboxes.

@danharrin hope the latest changes are what you were looking for.

File: packages/forms/resources/views/components/checkbox-list.blade.php [hunk 2]
```diff
@@ -108,5 +145,16 @@
...
+        @if ($isSearchable())
+            <div
+                x-cloak
+                x-show="visibleCheckboxListItems.length === 0"
+                class="filament-forms-checkbox-list-component-no-search-results-message text-sm"
+            >
+                {{ $getNoSearchResultsMessage() }}
+            </div>
+        @endif
+
```
Comment (zepfietje) [inline packages/forms/resources/views/components/checkbox-list.blade.php]:
```suggestion
```

File: packages/forms/resources/views/components/checkbox-list.blade.php [hunk 3]
```diff
@@ -74,7 +106,12 @@
         >
             @forelse ($getOptions() as $optionValue => $optionLabel)
                 <div wire:key="{{ $this->id }}.{{ $getStatePath() }}.{{ $field::class }}.options.{{ $optionValue }}">
-                    <label class="flex items-center space-x-3 rtl:space-x-reverse">
+                    <label
+                        class="rtl:space-x-reverse flex items-center space-x-3"
```
Comment (zepfietje) [inline packages/forms/resources/views/components/checkbox-list.blade.php]:
```suggestion
                        class="flex items-center space-x-3 rtl:space-x-reverse"
```

File: packages/forms/resources/views/components/checkbox-list.blade.php [hunk 4]
```diff
@@ -62,6 +84,16 @@
             @endif
         </div>
 
+        @if ($isSearchable())
+            <input
+                {!! $isDisabled() ? 'disabled' : null !!}
+                class="focus:border-primary-500 focus:ring-1 focus:ring-inset focus:ring-primary-500 disabled:opacity-70 block w-full transition duration-75 border-gray-300 rounded-lg shadow-sm mb-2"
```
Comment (zepfietje) [inline packages/forms/resources/views/components/checkbox-list.blade.php]:
```suggestion
                class="mb-2 block w-full rounded-lg border-gray-300 shadow-sm transition duration-75 focus:border-primary-500 focus:ring-1 focus:ring-inset focus:ring-primary-500 disabled:opacity-70"
```

File: packages/forms/resources/views/components/checkbox-list.blade.php [hunk 5]
```diff
@@ -62,6 +84,16 @@
             @endif
         </div>
 
+        @if ($isSearchable())
+            <input
+                {!! $isDisabled() ? 'disabled' : null !!}
```
Comment (zepfietje) [inline packages/forms/resources/views/components/checkbox-list.blade.php]:
```suggestion
                @disabled($isDisabled())
```

File: packages/forms/resources/views/components/checkbox-list.blade.php [hunk 6]
```diff
@@ -35,10 +39,28 @@
...
+                return checkboxListItem.querySelector('span').innerText.toLowerCase().includes(this.search.toLowerCase())
+            })
+        }
+
     }">
         <div wire:key="{{ $this->id }}.{{ $getStatePath() }}.{{ $field::class }}.wrapper">
             @if ($isBulkToggleable() && count($getOptions()))
-                <div x-cloak class="mb-2" wire:key="{{ $this->id }}.{{ $getStatePath() }}.{{ $field::class }}.buttons">
+                <div
+                    x-cloak
```
Comment (zepfietje) [inline packages/forms/resources/views/components/checkbox-list.blade.php:76]:
```suggestion
                    x-cloak=""
```

**Review by zepfietje — CHANGES_REQUESTED** [review]
Just a quick review for now.

File: packages/forms/resources/views/components/checkbox-list.blade.php (hunk 6 above)
Comment (danharrin) [reply packages/forms/resources/views/components/checkbox-list.blade.php:76]: This is only necessary when using x-cloak Blade components

File: packages/forms/resources/views/components/checkbox-list.blade.php (hunk 5 above)
Comment (danharrin) [reply packages/forms/resources/views/components/checkbox-list.blade.php]: Not L8 compatible

**Comment by zepfietje** [comment]
I'll still do a proper review of the functionality later. 👍

**Comment by danharrin** [comment]
thanks John!

File: packages/forms/resources/views/components/checkbox-list.blade.php [hunk 7]
```diff
@@ -13,32 +13,70 @@
...
+
+        search: '',
+
+        visibleCheckboxListOptions: [],
+
         init: function () {
             this.checkIfAllCheckboxesAreChecked()
 
             Livewire.hook('message.processed', () => {
                 this.checkIfAllCheckboxesAreChecked()
```
Comment (juliomotol) [inline packages/forms/resources/views/components/checkbox-list.blade.php:29]: Any particular reason why `this.updateVisibleCheckboxListOptions()` wasn't called here? We're having issue when we load the options dynamically. 

```php
Forms\Components\CheckboxList::make('states')
    ->options(
        fn (Closure $get) => ($country = $get('country'))
            ? State::whereCountryId($country)->pluck('name', 'id')
            : []
    )
```

It has been observed that the DOM is getting updated but the option are hidden.

Comment (zepfietje) [reply packages/forms/resources/views/components/checkbox-list.blade.php:29]: Feel free to PR a fix.

Comment (zepfietje) [reply packages/forms/resources/views/components/checkbox-list.blade.php:29]: @danharrin, might have to check if this still works with the Livewire hooks in v3.

**Pull request merged**